lms/djangoapps/grades/models.py

```diff
@@ -257,8 +257,14 @@ def save_grade(cls, **kwargs):
                     usage_key=usage_key,
                     defaults=kwargs,
                 )
+            log.info("persistent subsection grade successfully saved: {0}".format(unicode(grade)))
         except IntegrityError:
             cls.update_grade(user_id=user_id, usage_key=usage_key, **kwargs)
+            log.info("integrity error trying to save grade for user {0}, usage key {1}, defaults {2}".format(
```

Comment:
The `IntegrityError` handling exists here because of the default `REPEATABLE_READ` transaction level we're stuck with - it's possible for two requests to begin simultaneously with a "snapshot" of the database in which the grade has not been created. Whichever request is second will not be able to "see" the new grade created by the first, so it will try to create as well, leading to the `IntegrityError`.

Now that I'm looking at this again, I don't think this will work either - the `cls.update_grade` call will be just as unable to read the newly saved grade as the initial `get_or_create`, so it will fail too. I guess the best solution is to log and raise an error?

Preventing this scenario and allowing for retries may be an argument in favor of async grade updates.